### PR TITLE
Disable etcd scale down for shoot clusters with purpose "production".

### PR DESCRIPTION
**Pull request by amshuman-kr**
**How to categorize this PR?**

/area control-plane cost
/kind post-mortem

**What this PR does / why we need it**:

Disable etcd scale down for shoot clusters with purpose "production". This avoids multiple etcd restarts during the shoot's maintenance time window if VPA recommendation for scale down is inappropriate.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

cc @vlerenc @rfranzke @dguendisch 

**Release note**:

```other operator
Disable etcd scale down for shoot clusters with purpose "production". This avoids multiple etcd restarts during the shoot's maintenance time window if VPA recommendation for scale down is inappropriate.
```

### CODE REVIEW

**Comment by vlerenc** [comment]
Thank you @amshuman-kr !

**Comment by amshuman-kr** [comment]
The failing test case seems to be unrelated to the changes in this PR.
```
• Failure [26.150 seconds]
BackupBucket
/go/src/github.com/gardener/gardener/extensions/pkg/controller/backupbucket/backupbucket_test.go:55
  should successfully create and delete a BackupBucket (respecting operation annotation) [It]
  /go/src/github.com/gardener/gardener/extensions/pkg/controller/backupbucket/backupbucket_test.go:64

  Expected success, but got an error:
      <*errors.StatusError | 0xc0009f2280>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {
                  SelfLink: "",
                  ResourceVersion: "",
                  Continue: "",
                  RemainingItemCount: nil,
              },
              Status: "Failure",
              Message: "etcdserver: request timed out",
              Reason: "",
              Details: nil,
              Code: 500,
          },
      }
      etcdserver: request timed out

  /go/src/github.com/gardener/gardener/extensions/pkg/controller/backupbucket/backupbucket_test.go:200
------------------------------
STEP: running cleanup actions
STEP: delete backupbucket
STEP: delete secret
Failure [17.308 seconds]
[AfterSuite] AfterSuite 
/go/src/github.com/gardener/gardener/extensions/pkg/controller/backupbucket/backupbucket_suite_test.go:68

  Expected success, but got an error:
      <*errors.StatusError | 0xc00120e500>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {
                  SelfLink: "",
                  ResourceVersion: "",
                  Continue: "",
                  RemainingItemCount: nil,
              },
              Status: "Failure",
              Message: "secrets \"cloudprovider\" not found",
              Reason: "NotFound",
              Details: {
                  Name: "cloudprovider",
                  Group: "",
                  Kind: "secrets",
                  UID: "",
                  Causes: nil,
                  RetryAfterSeconds: 0,
              },
              Code: 404,
          },
      }
      secrets "cloudprovider" not found

  /go/src/github.com/gardener/gardener/extensions/pkg/controller/backupbucket/backupbucket_test.go:187
------------------------------


Summarizing 1 Failure:

[Fail] BackupBucket [It] should successfully create and delete a BackupBucket (respecting operation annotation) 
/go/src/github.com/gardener/gardener/extensions/pkg/controller/backupbucket/backupbucket_test.go:200

Ran 2 of 2 Specs in 132.375 seconds
FAIL! -- 1 Passed | 1 Failed | 0 Pending | 0 Skipped
```

File: pkg/operation/botanist/etcd.go [hunk 1]
```diff
@@ -102,6 +105,14 @@ func (b *Botanist) DeployEtcd(ctx context.Context) error {
 		})
 	}
 
+	if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeProduction {
+		var hvpaConfig = b.Shoot.Components.ControlPlane.EtcdMain.GetHVPAConfig()
+		if hvpaConfig != nil {
+			hvpaConfig.ScaleDownUpdateMode = pointer.StringPtr(hvpav1alpha1.UpdateModeOff)
+			b.Shoot.Components.ControlPlane.EtcdMain.SetHVPAConfig(hvpaConfig)
+		}
+	}
```
Comment (rfranzke) [inline pkg/operation/botanist/etcd.go]: Can you do this already in the `DefaultEtcd` function? Then you don't need to extend the interface with `GetHVPAConfig`?

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by ialidzhikov** [comment]
@amshuman-kr , could you please update `./docs/shoot_purposes.md` (the document that describes the behavioral differences between the Shoot purposes)?

**Comment by amshuman-kr** [comment]
@rfranzke @ialidzhikov I have made the changes you suggested. Can you PTAL?

File: pkg/operation/botanist/etcd.go [hunk 2]
```diff
@@ -43,7 +44,7 @@ import (
 var NewEtcd = etcd.New
 
 // DefaultEtcd returns a deployer for the etcd.
-func (b *Botanist) DefaultEtcd(role string, class etcd.Class) (etcd.Etcd, error) {
+func (b *Botanist) DefaultEtcd(role string, class etcd.Class, purpose gardencorev1beta1.ShootPurpose) (etcd.Etcd, error) {
```
Comment (rfranzke) [inline pkg/operation/botanist/etcd.go]:
```suggestion
func (b *Botanist) DefaultEtcd(role string, class etcd.Class) (etcd.Etcd, error) {
```

File: pkg/operation/botanist/etcd.go [hunk 3]
```diff
@@ -63,9 +64,15 @@ func (b *Botanist) DefaultEtcd(role string, class etcd.Class) (etcd.Etcd, error)
 	if b.ManagedSeed != nil {
 		hvpaEnabled = gardenletfeatures.FeatureGate.Enabled(features.HVPAForShootedSeed)
 	}
+
+	scaleDownUpdateMode := hvpav1alpha1.UpdateModeMaintenanceWindow
+	if class == etcd.ClassImportant && purpose == gardencorev1beta1.ShootPurposeProduction {
```
Comment (rfranzke) [inline pkg/operation/botanist/etcd.go]:
```suggestion
	if class == etcd.ClassImportant && b.Shoot.Purpose == gardencorev1beta1.ShootPurposeProduction {
```

File: pkg/operation/botanist/etcd.go [hunk 4]
```diff
@@ -63,9 +64,15 @@ func (b *Botanist) DefaultEtcd(role string, class etcd.Class) (etcd.Etcd, error)
...
+		scaleDownUpdateMode = hvpav1alpha1.UpdateModeOff
+	}
```
Comment (rfranzke) [inline pkg/operation/botanist/etcd.go:71]:
```suggestion
	}

```

File: pkg/operation/botanist/component/etcd/etcd.go [hunk 5]
```diff
@@ -358,11 +358,11 @@ func (e *etcd) Deploy(ctx context.Context) error {
...
 		var (
-			hpaLabels                   = map[string]string{v1beta1constants.LabelRole: "etcd-hpa-" + e.role}
-			vpaLabels                   = map[string]string{v1beta1constants.LabelRole: "etcd-vpa-" + e.role}
-			updateModeAuto              = hvpav1alpha1.UpdateModeAuto
-			updateModeMaintenanceWindow = hvpav1alpha1.UpdateModeMaintenanceWindow
-			containerPolicyOff          = autoscalingv1beta2.ContainerScalingModeOff
+			hpaLabels           = map[string]string{v1beta1constants.LabelRole: "etcd-hpa-" + e.role}
+			vpaLabels           = map[string]string{v1beta1constants.LabelRole: "etcd-vpa-" + e.role}
+			updateModeAuto      = hvpav1alpha1.UpdateModeAuto
+			scaleDownUpdateMode = e.hvpaConfig.ScaleDownUpdateMode
```
Comment (rfranzke) [inline pkg/operation/botanist/component/etcd/etcd.go]: Is this variable helpful? It seems `e.hvpaConfig.ScaleDownUpdateMode` can be used directly in line 427?

File: pkg/operation/botanist/botanist.go [hunk 6]
```diff
@@ -103,11 +103,11 @@ func New(ctx context.Context, o *operation.Operation) (*Botanist, error) {
 	}
 
 	// control plane components
-	o.Shoot.Components.ControlPlane.EtcdMain, err = b.DefaultEtcd(v1beta1constants.ETCDRoleMain, etcd.ClassImportant)
+	o.Shoot.Components.ControlPlane.EtcdMain, err = b.DefaultEtcd(v1beta1constants.ETCDRoleMain, etcd.ClassImportant, o.Shoot.Purpose)
```
Comment (rfranzke) [inline pkg/operation/botanist/botanist.go]:
```suggestion
	o.Shoot.Components.ControlPlane.EtcdMain, err = b.DefaultEtcd(v1beta1constants.ETCDRoleMain, etcd.ClassImportant)
```

File: pkg/operation/botanist/botanist.go [hunk 7]
```diff
@@ -103,11 +103,11 @@ func New(ctx context.Context, o *operation.Operation) (*Botanist, error) {
...
 	if err != nil {
 		return nil, err
 	}
-	o.Shoot.Components.ControlPlane.EtcdEvents, err = b.DefaultEtcd(v1beta1constants.ETCDRoleEvents, etcd.ClassNormal)
+	o.Shoot.Components.ControlPlane.EtcdEvents, err = b.DefaultEtcd(v1beta1constants.ETCDRoleEvents, etcd.ClassNormal, o.Shoot.Purpose)
```
Comment (rfranzke) [inline pkg/operation/botanist/botanist.go]:
```suggestion
	o.Shoot.Components.ControlPlane.EtcdEvents, err = b.DefaultEtcd(v1beta1constants.ETCDRoleEvents, etcd.ClassNormal)
```

**Comment by amshuman-kr** [comment]
/unhold

@rfranzke I have applied your suggestion, fixed the tests and squashed the commits. Can you please PTAL? Hopefully, one last time 😅

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**